core/block/object/objectcreator/smartblock.go

```diff
@@ -0,0 +1,150 @@
+		objectTypeKeys = []domain.TypeKey{bundle.TypeKeyPage}
+	}
+	sbType := objectTypeKeysToSmartBlockType(objectTypeKeys)
+
+	createState.SetDetailAndBundledRelation(bundle.RelationKeySpaceId, pbtypes.String(spc.Id()))
+
+	ev := &metrics.CreateObjectEvent{
+		SetDetailsMs: time.Since(startTime).Milliseconds(),
+	}
+
+	ctx = context.WithValue(ctx, eventCreate, ev)
+	initFunc := func(id string) *smartblock.InitContext {
+		createState.SetRootId(id)
+		return &smartblock.InitContext{
+			Ctx:            ctx,
+			ObjectTypeKeys: objectTypeKeys,
+			State:          createState,
+			RelationKeys:   generateRelationKeysFromState(createState),
+			SpaceID:        spc.Id(),
+		}
+	}
+
+	sb, err := createSmartBlock(ctx, spc, initFunc, createState, sbType)
+	if err != nil {
+		return "", nil, err
+	}
+
+	newDetails = sb.CombinedDetails()
+	id = sb.Id()
+
+	if sbType == coresb.SmartBlockTypeObjectType && pbtypes.GetInt64(newDetails, bundle.RelationKeyLastUsedDate.String()) == 0 {
+		objecttype.UpdateLastUsedDate(spc, s.objectStore, domain.TypeKey(
+			strings.TrimPrefix(pbtypes.GetString(newDetails, bundle.RelationKeyUniqueKey.String()), addr.ObjectTypeKeyToIdPrefix)),
+		)
+	} else if pbtypes.GetInt64(newDetails, bundle.RelationKeyOrigin.String()) == int64(model.ObjectOrigin_none) {
+		objecttype.UpdateLastUsedDate(spc, s.objectStore, objectTypeKeys[0])
+	}
+
+	ev.SmartblockCreateMs = time.Since(startTime).Milliseconds() - ev.SetDetailsMs - ev.WorkspaceCreateMs - ev.GetWorkspaceBlockWaitMs
+	ev.SmartblockType = int(sbType)
+	ev.ObjectId = id
+	metrics.SharedClient.RecordEvent(*ev)
+	return id, newDetails, nil
+}
+
+func objectTypeKeysToSmartBlockType(typeKeys []domain.TypeKey) coresb.SmartBlockType {
+	// TODO Add validation for types that user can't create
+
+	if slices.Contains(typeKeys, bundle.TypeKeyTemplate) {
+		return coresb.SmartBlockTypeTemplate
+	}
+	typeKey := typeKeys[0]
+
+	switch typeKey {
+	case bundle.TypeKeyObjectType:
+		return coresb.SmartBlockTypeObjectType
+	case bundle.TypeKeyRelation:
+		return coresb.SmartBlockTypeRelation
+	case bundle.TypeKeyRelationOption:
+		return coresb.SmartBlockTypeRelationOption
+	default:
+		return coresb.SmartBlockTypePage
+	}
+}
+
+func createSmartBlock(
+	ctx context.Context, spc clientspace.Space, initFunc objectcache.InitFunc, st *state.State, sbType coresb.SmartBlockType,
+) (smartblock.SmartBlock, error) {
+	if uKey := st.UniqueKeyInternal(); uKey != "" {
+		uk, err := domain.NewUniqueKey(sbType, uKey)
+		if err != nil {
+			return nil, err
+		}
+		return spc.DeriveTreeObject(ctx, objectcache.TreeDerivationParams{
+			Key:      uk,
+			InitFunc: initFunc,
+		})
+	}
+	return spc.CreateTreeObject(ctx, objectcache.TreeCreationParams{
+		Time:           time.Now(),
+		SmartblockType: sbType,
+		InitFunc:       initFunc,
+	})
+}
+
+func generateRelationKeysFromState(st *state.State) (relationKeys []string) {
+	if st == nil {
+		return
+	}
+	details := st.CombinedDetails().GetFields()
```

Comment:
CombinedDetails makes unnecessary allocations. Just iterate details and local details separately

Reply:
Done!